core/src/unique.rs

```diff
@@ -0,0 +1,338 @@
+        where F: FnOnce() -> Fut,
+              T: Clone,
+              Fut: IntoFuture<Item = (), Error = IoError>,
+              Fut::Future: Send + Sync + 'static, // TODO: 'static :-/
+    {
+        match &mut *self.inner.lock() {
+            UniqueConnecInner::Empty => (),
+            _ => return UniqueConnecFuture { inner: Arc::downgrade(&self.inner) },
+        };
+
+        // The mutex is unlocked when we call `or`, in order to avoid potential deadlocks.
+        let dial_fut = or().into_future();
+
+        let mut inner = self.inner.lock();
+        // Since we unlocked the mutex, it's possible that the object was filled in the meanwhile.
+        // Therefore we check again whether it's still `Empty`.
+        if let UniqueConnecInner::Empty = &mut *inner {
+            *inner = UniqueConnecInner::Pending {
+                tasks_waiting: Vec::new(),
+                dial_fut: Box::new(dial_fut),
+            };
+        }
+
+        UniqueConnecFuture { inner: Arc::downgrade(&self.inner) }
+    }
+
+    /// Puts `value` inside the object. The second parameter is a future whose completion will
+    /// clear up the content. Returns an adjusted version of that same future.
+    ///
+    /// If `clear()` is called, the returned future will automatically complete with an error.
+    ///
+    /// Has no effect if the object already contains something.
+    pub fn set_until<F>(&self, value: T, until: F) -> impl Future<Item = (), Error = F::Error>
+        where F: Future<Item = ()>
+    {
+        let mut tasks_to_notify = Vec::new();
+
+        let mut inner = self.inner.lock();
+        let (on_clear, on_clear_rx) = oneshot::channel();
+        match mem::replace(&mut *inner, UniqueConnecInner::Full { value, on_clear }) {
+            UniqueConnecInner::Empty => {},
+            UniqueConnecInner::Errored(_) => {},
+            UniqueConnecInner::Pending { tasks_waiting, .. } => {
+                tasks_to_notify = tasks_waiting;
+            },
+            old @ UniqueConnecInner::Full { .. } => {
+                // Keep the old value.
+                *inner = old;
+            },
+        };
+        drop(inner);
+
+        // The mutex is unlocked when we notify the pending tasks.
+        for task in tasks_to_notify {
+            task.notify();
+        }
+
+        let inner_clone = self.inner.clone();
+        let mut called_once = false;
+        until
+            .select(on_clear_rx.then(|_| Ok(())))
+            .map(|((), _)| ())
+            .map_err(|(err, _)| err)
+            .then(move |val| {
+                assert!(!called_once, "Future::poll() called again after returning Some");
+                called_once = true;
+                let mut inner = inner_clone.lock();
+                match mem::replace(&mut *inner, UniqueConnecInner::Empty) {
+                    UniqueConnecInner::Full { .. } => (),
+                    _ => panic!("Wrong state in the UniqueConnec ; programmer error")
+                }
```

Comment:
Maybe I am missing something, but in case `UniqueConnecInner` was already `Full`, `set_until` will stick with the previous value (line 162). The `until` future is run regardless though and calls to `clear` will not affect it (as `on_clear` from line 153 is gone), so with potentially two `until` futures running, the first would reset to `Empty` and the second panic here, no?

Reply:
I fixed this by directly returning `until` if the object is already full.
However I also removed the `panic!`, because having an error or calling `clear()` would trigger it as well.